controllers/domachine_controller.go

```diff
@@ -0,0 +1,290 @@
+		return reconcile.Result{}, nil
+	}
+
+	// If the DOMachine doesn't have our finalizer, add it.
+	if !util.Contains(domachine.Finalizers, infrav1.MachineFinalizer) {
+		domachine.Finalizers = append(domachine.Finalizers, infrav1.MachineFinalizer)
+	}
+
+	if !machineScope.Cluster.Status.InfrastructureReady {
+		machineScope.Info("Cluster infrastructure is not ready yet")
+		return reconcile.Result{}, nil
+	}
+
+	// Make sure bootstrap data is available and populated.
+	if machineScope.Machine.Spec.Bootstrap.Data == nil {
+		machineScope.Info("Bootstrap data is not yet available")
+		return reconcile.Result{}, nil
+	}
+
+	computesvc := computes.NewService(ctx, clusterScope)
+	droplet, err := computesvc.GetDroplet(machineScope.GetInstanceID())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if droplet == nil {
+		droplet, err = computesvc.CreateDroplet(machineScope)
+		if err != nil {
+			errs := errors.Errorf("Failed to create droplet instance for DOMachine %s/%s: %v", domachine.Namespace, domachine.Name, err)
+			machineScope.SetErrorReason(capierrors.CreateMachineError)
+			machineScope.SetErrorMessage(errs)
+			r.Recorder.Event(domachine, corev1.EventTypeWarning, "InstanceCreatingError", errs.Error())
+			return reconcile.Result{}, errs
+		}
+		r.Recorder.Eventf(domachine, corev1.EventTypeNormal, "InstanceCreated", "Created new droplet instance - %s", droplet.Name)
+	}
+
+	machineScope.SetProviderID(strconv.Itoa(droplet.ID))
+	machineScope.SetInstanceStatus(infrav1.DOResourceStatus(droplet.Status))
+
+	addrs, err := computesvc.GetDropletAddress(droplet)
+	if err != nil {
+		machineScope.SetErrorMessage(errors.New("failed to getting droplet address"))
+		return reconcile.Result{}, err
+	}
+	machineScope.SetAddresses(addrs)
+
+	// Proceed to reconcile the DOMachine state.
+	switch infrav1.DOResourceStatus(droplet.Status) {
+	case infrav1.DOResourceStatusNew:
+		machineScope.Info("Machine instance is pending", "instance-id", machineScope.GetInstanceID())
+		return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+	case infrav1.DOResourceStatusRunning:
+		machineScope.Info("Machine instance is active", "instance-id", machineScope.GetInstanceID())
+		machineScope.SetReady()
+		r.Recorder.Eventf(domachine, corev1.EventTypeNormal, "DOMachineReady", "DOMachine %s - has ready status", droplet.Name)
+		return reconcile.Result{}, nil
+	default:
+		machineScope.SetErrorReason(capierrors.UpdateMachineError)
+		machineScope.SetErrorMessage(errors.Errorf("Instance status %q is unexpected", droplet.Status))
+		return reconcile.Result{}, nil
+	}
+}
+
+func (r *DOMachineReconciler) reconcileDelete(ctx context.Context, machineScope *scope.MachineScope, clusterScope *scope.ClusterScope) (_ reconcile.Result, reterr error) {
+	machineScope.Info("Reconciling delete DOMachine")
+	domachine := machineScope.DOMachine
+
+	defer func() {
+		if reterr == nil {
+			domachine.Finalizers = util.Filter(domachine.Finalizers, infrav1.MachineFinalizer)
+		}
+	}()
```

Comment:
It looks like everything returns an error here with the exception of the final return on line 289. I'd suggest avoiding the defer here and just removing the finalizer right before that return.

Reply:
already updated @detiber